sound/soc/sof/ipc4-pcm.c

```diff
@@ -586,11 +587,57 @@ static int sof_ipc4_pcm_get_time_info(struct snd_soc_component *component,
 	return 0;
 }
 
+static snd_pcm_sframes_t sof_ipc4_pcm_delay(struct snd_soc_component *component,
+					    struct snd_pcm_substream *substream)
+{
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct sof_ipc4_llp_reading_slot llp, last_llp;
+	struct snd_sof_pcm_stream *stream;
+	struct snd_sof_pcm *spcm;
+	snd_pcm_uframes_t pos;
+	int time = 100;
+	u64 link_pos;
+
+	spcm = snd_sof_find_spcm_dai(component, rtd);
+	if (!spcm)
+		return 0;
+
+	stream = &spcm->stream[substream->stream];
+
+	if (stream->llp_offset) {
+		/*
+		 * Host dma position is gotten from mmio in host space but llp counter
+		 * is updated by fw by each schedule time, 1ms in common case. To
+		 * synchronize them, host will read host dma position when the llp counter
+		 * is just changed by fw. Driver checks the llp change every 10 us, so the
+		 * error time will about 10 ~ 15 us.
+		 */
+		sof_mailbox_read(stream->sdev, stream->llp_offset, &last_llp, sizeof(last_llp));
+		while (time--) {
+			usleep_range(10, 15);
+
+			sof_mailbox_read(stream->sdev, stream->llp_offset, &llp, sizeof(llp));
+			if (llp.reading.llp_l != last_llp.reading.llp_l)
+				break;
+		}
+	}
+
+	pos = sof_ops(stream->sdev)->pcm_pointer(stream->sdev, substream);
```

Comment:
You are disposing the position reported by the pcm_pointer() callback?

Don't you need that sort of thing for a delay calculation?

Reply:
pos is removed